### PR TITLE
CORE-329: Add createOrUpdate endpoint + refactoring

### PR DESCRIPTION
**Pull request by philippemiguet**
- Create new createOrUpdate endpoint to move logic from buffer-login to the session-service
- Jumping on the occasion to clean up the repo :)

### CODE REVIEW

**Comment by philippemiguet** [comment]
Hey @esclapes, I would love your thoughts on this PR :) 

1. How does the signature of the `createOrUpdate` function look to you? I'm not too sure how to approach it actually.
2. It made me realize the `/create` endpoint doesn't require a version to be passed but always use the one from `process.env.SESSION_VERSION`: it feels like it's killing the purpose of versioning, maybe? 🤔 How do you understand versioning in the `session-service`?

File: src/index.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    .ping()
+    .then(() => res.status(200).json({ status: 'awesome' }))
+    .catch(() => res.status(500).json({ status: 'cannot reach redis' }));
+});
+
+app.use(apiError);
+
+module.exports.app = app;
+
+if (process.env.NODE_ENV !== 'test') {
```
[the remark this answers is not in the document]
Comment (philippemiguet) [reply src/index.js]: Hey @esclapes how much should I move to the `app.js, all [this section](https://github.com/bufferapp/session-service/blob/e048f65df481efed48b9d56b03aa0258a6bdc5b6/src/index.js#L13-L50)?

File: src/rpc/createOrUpdate.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+const create = require('./create');
+const update = require('./update');
+
+module.exports = async ({ token, session, sessionVersion, userId, accountId }) => {
+  if (token) {
+    return update({ token, session, sessionVersion });
+  }
+
+  return create({ session, userId, accountId });
+};
```
Comment (esclapes) [inline src/rpc/createOrUpdate.js]: I see that `createOrUpdate` method in the `session-manager` is using the `create` and `update` methods under the hood. Do we still want this RPC method here?

File: src/rpc/index.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+const { rpc, method } = require('@bufferapp/buffer-rpc');
+
+module.exports = rpc(
+  method('create', create),
+  method('createOrUpdate', createOrUpdate),
+  method('destroy', destroy),
+  method('get', get),
+  method('getActiveSessions', getActiveSessions),
+  method('update', update),
+);
```
Comment (esclapes) [inline src/rpc/index.js]: :heart_eyes: :heart_eyes:

File: Dockerfile [hunk 4]
```diff
@@ -8,8 +8,8 @@ ENV NODE_ENV production
 COPY package.json /usr/src/app
 RUN npm install
 
-COPY . /usr/src/app
+COPY src /usr/src/app/src
```
[the remark this answers is not in the document]
Comment (esclapes) [reply Dockerfile]: Thanks for the detailed explanation @philippemiguet . :100: for consistency!!

File: package.json [hunk 5]
```diff
@@ -2,10 +2,10 @@
   "name": "@bufferapp/session-service",
   "version": "0.0.1",
   "description": "Buffer session service",
-  "main": "index.js",
+  "main": "src/index.js",
   "scripts": {
-    "start": "micro",
-    "lint": "eslint . --ignore-pattern coverage node_modules",
+    "start": "node .",
```
[the remark this answers is not in the document]
Comment (esclapes) [reply package.json]: I love the great care you are putting into consistency @philippemiguet :taco:

File: src/rpc/createOrUpdate.js (hunk 2 above)
Comment (philippemiguet) [reply src/rpc/createOrUpdate.js]: Haha, great catch @esclapes, indeed we only one or the other change 😂 🤦‍♂️ 

Solutions are:
 - Keep the new endpoint `/createOrUpdate` in the session-service and make the logic in session-manager more straightforward and always call this endpoint.
 - keep the logic of the session manager to call either /create or /update, and remove this new endpoint.

@esclapes what do you think is best?

File: src/rpc/index.js (hunk 3 above)
Comment (philippemiguet) [reply src/rpc/index.js]: Took it from what you shared on the buffer-rpc issue, it's much cleaner indeed!

File: src/libs/utils.js [hunk 6]
```diff
@@ -1,9 +1,34 @@
 const { createError } = require('@bufferapp/buffer-rpc');
+const ObjectPath = require('object-path');
 
 module.exports = {};
```
[the remark this answers is not in the document]
Comment (philippemiguet) [reply src/libs/utils.js]: Hey @esclapes, let's move to const + module.exports at the end like we do everywhere else! I think Harrison was using module.exports b/c the file was small, but as it grows and functions needs to call other from the same file, it's better to use const.

**Pull request merged**